### ISSUE TITLE
Autocomplete does not quote `String` filepaths when using external completers

### ISSUE THREAD

**Issue by Beinsezii** [nushell/nushell#10690]
### Describe the bug

When using Tab or similar to autocomplete file names, it only quotes paths with special chars/spaces if the command is expecting a path-like object. If the command arg is a String this does not happen, forcing the user to manually go back and add quotes to filenames.

### How to reproduce

Firstly, this impacts all external completers from the cookbook _except_ `carapace`. Use this minimal config, lifted straight from https://www.nushell.sh/cookbook/external_completers.html#external-completers

```nushell
let fish_completer = {|spans|
    fish --command $'complete "--do-complete=($spans | str join " ")"'
    | from tsv --flexible --noheaders --no-infer
    | rename value description
}
$env.config.completions.external = { enable: true max_results: 100 completer: $fish_completer }
```

Then take a file with all the naughty characters like `Movie (2023).mkv` and type Mo\<TAB\> after various commands

Commands accepting pathlike objects auto-quote as expected:
```
> mv `Movie (2023).mkv`
> touch `Movie (2023).mkv`
```
Commands expecting String like `start` and all external commands do not appear to quote:
```
> start Movie (2023).mkv
> mpv Movie (2023).mkv
```
"Why not lead with a quote?" you ask. Also doesn't work. Backticks ``` ` ``` and double quotes `"` block tab-complete, while single quotes `'` are erased.
```
> mpv "Mo<tab>
NO RECORDS FOUND
> mpv 'Mo<tab>
> mpv Movie (2023).mkv
```

### Expected behavior

Ideally, nu could auto-quote funky filenames so I don't have to hammer CTRL+Left back to the start of the filename to add a quote myself. I have an extremely obtuse workaround at https://github.com/nushell/nushell/issues/10690#issuecomment-2299916524 but ideally I think `nu` could have a way of parsing backslash escapes on its own.

Otherwise, maybe we should at least update the cookbook and maybe regular book, as this is a large footgun for new users.

### Screenshots

I have a video demonstration at https://github.com/nushell/nushell/issues/10690#issuecomment-2727095656

### Configuration

| key                | value                                                            |
| ------------------ | ---------------------------------------------------------------- |
| version            | 0.102.0                                                          |
| major              | 0                                                                |
| minor              | 102                                                              |
| patch              | 0                                                                |
| branch             | makepkg                                                          |
| commit_hash        | 1aa2ed1947a0b891398558fcf4e4289849cc5a1d                         |
| build_os           | linux-x86_64                                                     |
| build_target       | x86_64-unknown-linux-gnu                                         |
| rust_version       | rustc 1.84.1 (e71f9a9a9 2025-01-27) (Arch Linux rust 1:1.84.1-1) |
| cargo_version      | cargo 1.84.1 (66221abde 2024-11-19)                              |
| build_time         | 2025-02-05 21:56:07 +00:00                                       |
| build_rust_channel | release                                                          |
| allocator          | mimalloc                                                         |
| features           | default, sqlite, trash                                           |
| installed_plugins  | bexpand 1.3.100, polars 0.100.0                                  |

**Comment by fdncred** [comment]
My research into this, that I posted on Discord, was that I was guessing that it worked this way because `start` takes `SyntaxShape::String` as a parameter whereas `touch` takes a `SyntaxShape::Filepath` as a parameter.

**Comment by weirdan** [comment]
Autocompletion for `start` and externals now quotes the filename (with either double quote or backtick).

**Comment by Beinsezii** [comment]
Was that committed recently? I don't see it.

**Comment by weirdan** [comment]
@Beinsezii I'm on 0.96.0 which is almost a month old.

**Comment by Beinsezii** [comment]
I'm on 0.96.1 and the behavior is completely unchanged from when I opened the issue.

```
touch `Movie (2023).mkv`

 # completes without quotes which causes command to fail
start M<TAB>
start 'M<TAB>

# Doesn't complete at all
start "M<TAB>
start `M<TAB>
```

**Comment by weirdan** [comment]
Works for me on 0.96.1 as well: https://asciinema.org/a/2acDz8fKLFmQkZz7ss1Pxi1LN

**Comment by fdncred** [comment]
I can complete with single, double, or backtick quotes in this use case on macos in 0.96.2

**Comment by Beinsezii** [comment]
Was this a new config option somewhere? Completion works now under run0 but not my regular user.

**Comment by weirdan** [comment]
Here's my comletion options, just in case:
```console
$ $env.config.completions
╭────────────────┬──────────────────────────────────╮
│ case_sensitive │ false                            │
│ quick          │ true                             │
│ partial        │ true                             │
│ algorithm      │ prefix                           │
│                │ ╭─────────────┬────────────────╮ │
│ external       │ │ enable      │ true           │ │
│                │ │ max_results │ 100            │ │
│                │ │ completer   │ <Closure 1302> │ │
│                │ ╰─────────────┴────────────────╯ │
│ use_ls_colors  │ true                             │
╰────────────────┴──────────────────────────────────╯
```

**Comment by Beinsezii** [comment]
I found it. It seems if you set use the `fish` completer shown in the nu book
https://www.nushell.sh/cookbook/external_completers.html#fish-completer
String completions for files just kinda break.

**Comment by Beinsezii** [comment]
Alright so I see the problem.

The fish example directly uses plaintext results as the completion table which means files never get escaped.

Fish *has* an `--escape` flag but since nushell can't use backslash escapes you have to hack around it with something like
```nushell
let fish_completer = {|spans|
    fish --command $'complete --escape "--do-complete=($spans | str join " ")"'
    | $"value(char tab)description(char newline)" + $in
    | from tsv --flexible --no-infer
    | each {|i|
        if '\' in $i.value {
            $i | merge {'value': $"\"($i.value | str replace -a '\' '')\""}
        } else {$i}
    }
}
```
To crudely unescape and quote the values, but that's probably going to fail with edge cases.

If `fish complete` is syntactically incompatible with nushell's string system, I don't think it should be recommended in the book unless a more proper parsing strategy is introduced.

**Comment by r-darwish** [comment]
Maybe related, but this seems to happen with custom completions even if the fish completer is used:

```nushell
def _az_subscriptions [] {
    az account list | from json | each {|s| {value: $s.name, description: $s.tenantDisplayName}} 
}

def "az set-subscription" [subscription: string@_az_subscriptions] {
    az account set --subscription $subscription
}
```

I'm using this function and completions with spaces are not quoted, even if I change `subscription`'s type to `path`

**Comment by juchiast** [comment]
This only happens when I use the fish external completer, I this we should remove fish from this page https://www.nushell.sh/book/custom_completions.html or at least put a warning about this.

**Comment by youwen5** [comment]
The fish completions bring high enough utility that I don't think we should remove it as keeping my fish completions was a deciding factor in my decision to switch to Nushell. We should probably add a warning though.

**Comment by Beinsezii** [comment]
This can really happen to any completer since nu relies on the completer itself to do the escapes/quotes, and currently there's one in the entire world that supports nu which is carapace. All of the rest will need to have some sort of post processing done, like checking if the completion is a file or converting backslash escapes into nu escapes. Or maybe you could have carapace as the first layer then failover to fish/others if carapace is empty.

**Comment by blindFS** [comment]
@Beinsezii no longer the case in nu 0.102 on my machine, could please help to confirm?

**Comment by Beinsezii** [comment]
@blindFS What do you mean? Nothing has changed in nushell regarding this. I see they did move all the completes except carapace out of the book and into the cookbook, but there's still no notes about this behavior outside of this issue tracker.

If you think that's enough to close it then sure, but I wouldn't be surprised if it pops back up again when someone follows the cookbook then wonders why the built-in command `start` can't tab complete anymore.

**Comment by blindFS** [comment]
@Beinsezii 

https://github.com/user-attachments/assets/e4594533-f692-4445-9916-2c529825fb8c

Is this what you're expecting, or do you mean something else?

**Comment by Beinsezii** [comment]
@blindFS Yes, that's what should happen. You must be using Carapace. Use any of the other closures from the cookbook and it will fail to complete properly.

**Comment by blindFS** [comment]
> @blindFS Yes, that's what should happen. You must be using Carapace. Use any of the other closures from the cookbook and it will fail to complete properly.

Nope, start is a builtin command, nothing to do with the external completer here. You may want to recheck  your nu version and how it behaves without your config.

**Comment by Beinsezii** [comment]
External completers affect builtin commands that use the `string` data type. Using your exact config

`nu --version | wl-copy`
`0.102.0`

`$env.config.completions | to md | wl-copy`
|sort|case_sensitive|quick|partial|algorithm|external|use_ls_colors|
|-|-|-|-|-|-|-|
|smart|false|true|true|prefix|{enable: true, max_results: 100, completer: closure_430}|true|

any completer that _isn't_ carapace, `start` will fail to complete correctly for nushell.

**Comment by Beinsezii** [comment]
In this case `start` takes `path` as a string instead of `Path` so it works on general uri schemes, which is why `fish` is called instead of nu's built-in.

**Comment by Beinsezii** [comment]
https://github.com/user-attachments/assets/59d4b1eb-c883-4497-a4df-b9ceff76e6a0

**Comment by blindFS** [comment]
@Beinsezii first of all, you should be aware that nushell completions treat `start` and `mpv` differently (internal vs external). And the file completion shown in my screenshot is the fallback behavior for internal arguments of type path/string/..., as well as when the external completer returns nil.

**Comment by Beinsezii** [comment]
It treats `start` and `mpv` the exact same.

**Comment by blindFS** [comment]
> It treats `start` and `mpv` the exact same, as they both use general `string` completion

Could you please provide a minimal reproducible config file then?

**Comment by Beinsezii** [comment]
Try it yourself

```
# From https://www.nushell.sh/cookbook/external_completers.html#external-completers
let fish_completer = {|spans|
    fish --command $'complete "--do-complete=($spans | str join " ")"'
    | from tsv --flexible --noheaders --no-infer
    | rename value description
}
$env.config.completions.external = { enable: true max_results: 100 completer: $fish_completer }
```

**Comment by Beinsezii** [comment]
I have no idea how much more clear I can make it that `start` and `mpv` are completed the same way because they both use the string data type.

**Comment by blindFS** [comment]
> Try it yourself
> 
> ```
> # From https://www.nushell.sh/cookbook/external_completers.html#external-completers
> let fish_completer = {|spans|
>     fish --command $'complete "--do-complete=($spans | str join " ")"'
>     | from tsv --flexible --noheaders --no-infer
>     | rename value description
> }
> $env.config.completions.external = { enable: true max_results: 100 completer: $fish_completer }
> ```

I don't understand how could you be so rude for someone trying to solve your problem?

**Comment by blindFS** [comment]
> I have no idea how much more clear I can make it that `start` and `mpv` are completed the same way because they both use the string data type.

How could you be so confident about something you don't fully understand?

**Comment by Beinsezii** [comment]
> I don't understand how could you be so rude for someone trying to solve your problem?

I personally feel like you haven't been respecting the information I've provided, counterclaiming me multiple times

> Nope, start is a builtin command, nothing to do with the external completer here. You may want to recheck your nu version and how it behaves without your config.
> first of all, you should be aware that nushell completions treat `start` and `mpv` differently (internal vs external). [...]

Without having tried it for yourself. You seem to be approaching this with the mindset that it *has* to be a user error.

You can make a fresh nushell environment, set the config to exactly what I've provided, and experience firsthand what I've shown in my video recording.

I believe in `nushell` and want it to be the best it can be, so footguns like tab completions breaking is something I want to help resolve in some way.



> How could you be so confident about something you don't fully understand?

I feel like that's a reasonable interpretation given @fdncred 's comment https://github.com/nushell/nushell/issues/10690#issuecomment-1758425564
and the observations I've provided in video and code example. Even if they're not literally the same, they're functionally the same for the purpose of this ticket.

**Comment by blindFS** [comment]
@Beinsezii first of all, when I kindly asked you to help confirm whether it is fixed in nu 102. You acted like "wtf, can't you do it yourself?"

The true story is it at least works as expected with default config. You should update the reproduce steps at least.

**Comment by Beinsezii** [comment]
> The true story is it at least works as expected with default config. You should update the reproduce steps at least.

Yeah, I can update the top level that the issues with `start` and `mpv` are specific to non-carapace external completers. It's been a year and a half since I opened this, so I just forgot.


> first of all, when I kindly asked you to help confirm whether it is fixed in nu 102. You acted like "wtf, can't you do it yourself?"

I have no clue what you're talking about. My exact words

>  What do you mean? Nothing has changed in nushell regarding this.

Was simply me expressing that the issue is still the exact same in the latest tagged release as it was in 0.85. I'm not sure if this is a language barrier or what have you, but from my point of view I expressed confusion about the issue being supposedly resolved, then we disagreed on `start` using external completers, and now I'm being called names.

**Comment by blindFS** [comment]
> Was simply me expressing that the issue is still the exact same in the latest tagged release as it was in 0.85. I'm not sure if this is a language barrier or what have you, but from my point of view I expressed confusion about the issue being supposedly resolved, then we disagreed on `start` using external completers, and now I'm being called names.

I'm sick of babysitting you and teaching you how to respond properly. I'll leave this issue to someone else.

**Comment by fdncred** [comment]
If I start `nu -n`, then add the fish_completer as described in the description, and then try and repo this. This is what happens.
1. `touch mo<tab>`, completes like ``` touch `Movie (2023).mkv` ```
2. `mv mo<tab>`, completes like ``` mv `Movie (2023).mkv` ```
3. `start mo<tab>`, completes like ``` start `Movie (2023).mkv` ```
4. `mpv mo<tab>`, completes like ``` mpv Movie (2023).mkv ``` (which is wrong)

So, it appears that completions have gotten better but for some reason mpv still doesn't complete the same as the internal commands. I suspect that's because it is the only one going through the external completer. You can also repro this with external mv like `^mv mo<tab>`.

**Comment by blindFS** [comment]
> If I start `nu -n`, then add the fish_completer as described in the description, and then try and repo this. This is what happens.
> 
> 1. `touch mo<tab>`, completes like `` touch `Movie (2023).mkv` ``
> 2. `mv mo<tab>`, completes like `` mv `Movie (2023).mkv` ``
> 3. `start mo<tab>`, completes like `` start `Movie (2023).mkv` ``
> 4. `mpv mo<tab>`, completes like `mpv Movie (2023).mkv` (which is wrong)

Thanks for your attention, I believe that's what nu 103 will behave like. In nu 102, I can reproduce the issue for `start`.

> So, it appears that completions have gotten better but for some reason mpv still doesn't complete the same as the internal commands. I suspect that's because it is the only one going through the external completer. You can also repro this with external mv like `^mv mo<tab>`.

To me, this is more of a document issue, we can't control the behavior of user defined external completer closures. I think we should open another issue on the github.io repo.

**Comment by Beinsezii** [comment]
> Thanks for your attention, I believe that's what nu 103 will behave like. In nu 102, I can reproduce the issue for `start`.

I can confirm on the `master` branch with `~/.cargo/bin/nu -n` that `start` does behave properly now.

As for external commands, I was thinking short term documenting the footgun would be great.

Long term I would be curious about the viability of something like `str unescape` or `str nu-escape` to convert

`Movie\ \(2023\).mkv` -> ``` `Movie (2023).mkv` ```

As that would allow a cleaner backwards-compat with `sh` style escapes from other completers

**Comment by fdncred** [comment]
> Thanks for your attention, I believe that's what nu 103 will behave like. 

Agreed

> To me, this is more of a document issue

Agreed

> I was thinking short term documenting the footgun would be great

Agreed. Do you have time to do that?

I have no problem with better documentation.

> Long term  ...

I understand your example for `str unescape` I just wonder if there's some other way to do that? It kind of seems like a niche-single-task command. I mean, what else could it be used for? I guess it's kind of just a posix escape translator, which kind of makes me think we'd need a to-sh-escape and from-sh-escape type of thing. I'm just not sure how useful that would be.

**Comment by Beinsezii** [comment]
I'm on 103 now, and all of the completion problems have been resolved except the external command + external completer issue discussed.

> Agreed. Do you have time to do that?
> 
> I have no problem with better documentation.

PR open https://github.com/nushell/nushell.github.io/pull/1844

> > Long term  ...
> 
> I understand your example for `str unescape` I just wonder if there's some other way to do that? It kind of seems like a niche-single-task command. I mean, what else could it be used for? I guess it's kind of just a posix escape translator, which kind of makes me think we'd need a to-sh-escape and from-sh-escape type of thing. I'm just not sure how useful that would be.

I have some pure nu workarounds in the PR but none of them are really great IMO

**Comment by Beinsezii** [comment]
With 103 out and the docs having been updated https://github.com/nushell/nushell.github.io/pull/1844 I think this is certified good enough™
